src/algorithms/synthesis/dd_synthesis.cpp

```diff
@@ -259,23 +297,36 @@ namespace syrec {
         controlNonRoot(current, ctrlNonRoot, ctrlVec);
 
         TruthTable::Cube::Vector rootSigVec;
-        pathFromSrcDst(src, current, rootSigVec);
+        pathFromSrcDst(src, current.p, rootSigVec);
 
         const auto nQubits = static_cast<dd::QubitCount>(src.p->v + 1);
-        for (auto const& rootVec: rootSigVec) {
+
+        const auto rootSolution = esopOptimization(rootSigVec);
+
+        if (rootSolution.empty()) {
+            ctrlNonRoot.emplace(dd::Control{current.p->v, dd::Control::Type::pos});
+            for (std::size_t i = 0; i < targetVec.size(); i++) {
+                if (targetVec[i].has_value() && *(targetVec[i])) {
+                    applyOperation(nQubits, static_cast<dd::Qubit>(static_cast<std::size_t>(current.p->v) - (i + 1U)), src, ctrlNonRoot, dd);
+                }
+            }
+        }
+
+        for (auto const& rootVec: rootSolution) {
             dd::Controls ctrlFinal;
-            controlRoot(current, ctrlFinal, rootVec);
+
+            controlRootEsop(current, ctrlFinal, rootVec, rootSigVec[0].size());
+
             ctrlFinal.emplace(dd::Control{current.p->v, dd::Control::Type::pos});
+
             ctrlFinal.insert(ctrlNonRoot.begin(), ctrlNonRoot.end());
 
-            const auto targetSize = targetVec.size();
-            for (auto i = 0U; i < targetSize; ++i) {
+            for (std::size_t i = 0; i < targetVec.size(); i++) {
```

Comment:
This change should be reverted. Calculating the loop termination condition outside the loop is always preferable to re-compute it in every loop iteration. (Although in cases like these the compiler is probably clever enough to figure that out itself)